lib/auth/trust/trustv1/service.go

```diff
@@ -203,6 +206,79 @@ func (s *Service) UpsertCertAuthority(ctx context.Context, req *trustpb.UpsertCe
 	return req.CertAuthority, nil
 }
 
+// RotateExternalCertAuthority rotates external certificate authority,
+// this method is called by remote trusted cluster and is used to update
+// only public keys and certificates of the certificate authority.
+func (s *Service) RotateExternalCertAuthority(ctx context.Context, req *trustpb.RotateExternalCertAuthorityRequest) (*trustpb.RotateExternalCertAuthorityResponse, error) {
+	if req.CertAuthority == nil {
+		return nil, trace.BadParameter("missing certificate authority")
+	}
+
+	if err := services.ValidateCertAuthority(req.CertAuthority); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	authzCtx, err := authz.AuthorizeResourceWithVerbs(ctx, s.logger, s.authorizer, false, req.CertAuthority, types.VerbRotate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authz.AuthorizeAdminAction(ctx, authzCtx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName, err := s.authServer.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// this is just an extra precaution against local admins,
+	// because this is additionally enforced by RBAC as well
+	if req.CertAuthority.GetClusterName() == clusterName.GetClusterName() {
+		return nil, trace.BadParameter("can not rotate local certificate authority")
+	}
+
+	existing, err := s.cache.GetCertAuthority(ctx, types.CertAuthID{
+		Type:       req.CertAuthority.GetType(),
+		DomainName: req.CertAuthority.GetClusterName(),
+	}, true)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	updated := existing.Clone()
+	if err := updated.SetActiveKeys(req.CertAuthority.GetActiveKeys().Clone()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := updated.SetAdditionalTrustedKeys(req.CertAuthority.GetAdditionalTrustedKeys().Clone()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// a rotation state of "" gets stored as "standby" after
+	// CheckAndSetDefaults, so if `ca` came in with a zeroed rotation we must do
+	// this before checking if `updated` is the same as `existing` or the check
+	// will fail for no reason (CheckAndSetDefaults is idempotent, so it's fine
+	// to call it both here and in CompareAndSwapCertAuthority)
+	updated.SetRotation(req.CertAuthority.GetRotation())
+	if err := services.CheckAndSetDefaults(updated); err != nil {
```

Comment:
Is the indirection necessary here? Could we do:

```suggestion
	if err := updated.CheckAndSetDefaults(); err != nil {
```

If we can that is best, as it has stronger typing and works better with symbol search.

Reply:
`CheckAndSetDefaults` was removed from the `types.Resource` interface - https://github.com/gravitational/teleport/pull/35561 - so I'll have to do a type check either way.

Reply:
I thought the CertAuthoriry type carried it still.

Reply:
`CertAuthorityV2` does, I can add it to the `types.CertAuthority` if we want

Reply:
No strong opinions in this case.